File: evd_ros_backend/evd_ros_core/src/evd_script/program/machine_operations/machine_primitive.py
'''
Machine Primitive extends primitive to simplify the later definition of specific machine primitives.

A machine primitive is a way to generalize machine behavior in EvD. The actual behavior needs to be
supplied externally and is hooked into the EvD program runner.
'''

from ..primitive import Primitive


class MachinePrimitive(Primitive):

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'machine-primitive' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return Primitive.full_type_string() + cls.type_string()

    def __init__(self, machineUuid=None, type='', name='', uuid=None, parent=None, append_type=True, editable=True, deleteable=True):
        self._machine_uuid = None

        super(MachinePrimitive,self).__init__(
            type=MachinePrimitive.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type,
            editable=editable,
            deleteable=deleteable)

        self.machine_uuid = machineUuid

    def to_dct(self):
        msg = super(MachinePrimitive,self).to_dct()
        msg.update({
            'machine_uuid': self.machine_uuid
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        return cls(
            name=dct['name'],
            uuid=dct['uuid'],
            type=dct['type'],
            append_type=False,
            machineUuid=dct['machine_uuid'])

    '''
    Data accessor/modifier methods
    '''

    @property
    def machine_uuid(self):
        return self._machine_uuid

    @machine_uuid.setter
    def machine_uuid(self, value):
        if self._machine_uuid != value:
            self._machine_uuid = value
            self.updated_attribute('machine_uuid','set')

    def set(self, dct):
        if 'machine_uuid' in dct.keys():
            self.machine_uuid = dct['machine_uuid']

        super(MachinePrimitive,self).set(dct)

    '''
    Update Methods
    '''

    def deep_update(self):
        super(MachinePrimitive,self).deep_update()

        self.updated_attribute('machine_uuid','update')

    def shallow_update(self):
        super(MachinePrimitive,self).shallow_update()

        self.updated_attribute('machine_uuid','update')

File: evd_authoring_app/src/model/evd_script/program/machine_operations/machine_primitive.js
import { Primitive } from '../primitive';

export class MachinePrimitive extends Primitive {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'machine-primitive';
    }

    static fullTypeString() {
        return Primitive.fullTypeString() + MachinePrimitive.typeString();
    }

    constructor(machineUuid=null, type='', name='', uuid=null, parent=null, appendType=true) {
        this._machineUuid = null;

        super(
            type= (appendType) ? 'machine-primitive.'+type : type,
            name= name,
            uuid= uuid,
            parent= parent,
            appendType= appendType
        );

        this.machineUuid = machineUuid;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            machine_uuid: this.machineUuid
        };
        return msg;
    }

    static fromDict(dct) {
        return new MachinePrimitive(
            machineUuid= dct.machine_uuid,
            type= dct.type,
            name= dct.name,
            uuid= dct.uuid,
            appendType= false
        );
    }

    /*
    * Data accessor/modifier methods
    */

    get machineUuid() {
        return this._machineUuid;
    }

    set machineUuid(value) {
        if (this._machineUuid !== value) {
            this._machineUuid = value;
            this.updatedAttribute('machine_uuid','set');
        }
    }

    set(dct) {

        if ('machine_uuid' in dct) {
            this.machineUuid = dct.machine_uuid;
        }

        super.set(dct);
    }

    /*
    * Update methods
    */

    deepUpdate() {
        super.deepUpdate();

        this.updatedAttribute('machine_uuid','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('machine_uuid','update');
    }
}